File: modules/expo-mlkit-face-detection/ios/ExpoMLKitFaceDetectionModule.swift

import ExpoModulesCore
import MLKitFaceDetection
import ExpoMLKitCore

// Function to reject a promise with a specified message and domain
func rejectPromiseWithMessage(promise: Promise, message: String, domain: String) {
    promise.reject(
        NSError(domain: domain, code: 1, userInfo: [NSLocalizedDescriptionKey: message])
    )
}

let ERROR_DOMAIN: String = "red.infinite.expomlkit.FaceDetectorErrorDomain"


public class ExpoMLKitFaceDetectionModule: Module {
    var faceDetector: ExpoMLKitFaceDetector? = nil

    public func definition() -> ModuleDefinition {
        Name("ExpoMLKitFaceDetection")

        AsyncFunction("initialize") { (record: ExpoMLKitFaceDetectorOptionsRecord, promise: Promise) in
            let options = ExpoMLKitFaceDetectorOptions(record: record)
            self.faceDetector = ExpoMLKitFaceDetector(options: options)
            promise.resolve()
        }

        AsyncFunction("detectFaces") { (imagePath: String, promise: Promise) in
            let logger = Logger()

            Task {
                do {
                    guard let faceDetector = self.faceDetector else {
                        rejectPromiseWithMessage(promise: promise, message: "[ExpoMLKitFaceDetection.detectFace] Face detector not initialized", domain: ERROR_DOMAIN)
                        return
                    }
                    let image = try ExpoMLKitImage(imagePath: imagePath)
                    let result = try await faceDetector.detectFaces(image: image)
                    logger.debug(result)
                    // Use result to resolve promise
                    promise.resolve(result.record)

                } catch {
                    rejectPromiseWithMessage(promise: promise, message: "[ExpoMLKitFaceDetection.detectFace] Error Detecting Faces: \(error)", domain: ERROR_DOMAIN)
                }
            }
        }
    }
}

File: modules/expo-mlkit-face-detection/android/src/main/java/red/infinite/expomlkit/facedetection/ExpoMLKitFaceDetectionModule.kt
package red.infinite.expomlkit.facedetection

import ExpoMLKitFaceDetector
import ExpoMLKitFaceDetectorOptions
import ExpoMLKitFaceDetectorOptionsRecord
import android.net.Uri
import android.util.Log
import expo.modules.kotlin.Promise
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import kotlinx.coroutines.runBlocking


class ExpoMLKitFaceDetectionModule : Module() {


    private var faceDetector: ExpoMLKitFaceDetector? = null

    // Each module class must implement the definition function. The definition consists of components
    // that describes the module's functionality and behavior.
    // See https://docs.expo.dev/modules/module-api for more details about available components.
    override fun definition() = ModuleDefinition {
        // Sets the name of the module that JavaScript code will use to refer to the module. Takes a string as an argument.
        // Can be inferred from module's class name, but it's recommended to set it explicitly for clarity.
        // The module will be accessible from `requireNativeModule('ExpoMLKitFaceDetection')` in JavaScript.
        Name("ExpoMLKitFaceDetection")

        AsyncFunction("initialize") { options: ExpoMLKitFaceDetectorOptionsRecord, promise: Promise ->
            Log.d("ExpoMLKitFaceDetection", "initialize: Initializing Face Detection")
            faceDetector = ExpoMLKitFaceDetector(ExpoMLKitFaceDetectorOptions(options))
            Log.d("ExpoMLKitFaceDetection", "initialize: Initializing Face Detection")
            promise.resolve(null)
        }

        AsyncFunction("detectFaces") { imagePath: String, promise: Promise ->
            val imageUri = Uri.parse(imagePath)

            runBlocking suspend@{
                val faceDetector = faceDetector
                if (faceDetector == null) {
                    promise.reject(CodedException("ExpoMLKitFaceDetection: Face Detector not initialized"))
                    return@suspend
                }
                try {

                    var result =
                        faceDetector.detectFaces(imagePath, appContext).getOrElse {
                            promise.reject(CodedException("ExpoMLKitFaceDetection", it))
                            throw it
                        }

                    promise.resolve(result.record)

                } catch (e: Exception) {
                    promise.reject(CodedException("ExpoMLKitFaceDetection", e))
                }

            }
        }
    }
}
